src/category_theory/limits/mono_coprod.lean

```diff
@@ -0,0 +1,114 @@
+/-
+Copyright (c) 2022 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+
+import category_theory.limits.types
+import category_theory.morphism_property
+
+/-!
+
+# Categories where inclusions into coproducts are monomorphisms
+
+If `C` is a category that has finite coproducts, the class `mono_coprod C`
+expresses that left inclusions `A ⟶ A ⨿ B` are monomorphisms. If it is so,
+it is shown that right inclusions are also monomorphisms.
+
+TODO @joelriou: show that if `X : I → C` and `ι : J → I` is an injective map,
+then the canonical morphism `∐ (X ∘ ι) ⟶ ∐ X` is a monomorphism.
+
+-/
+
+noncomputable theory
+
+open category_theory category_theory.category category_theory.limits
+
+universe u
+
+namespace category_theory
+
+namespace limits
+
+variables (C : Type*) [category C] [hC : has_finite_coproducts C]
+
+include hC
+
+/-- This condition expresses that inclusion morphisms into coproducts are monomorphisms. -/
+class mono_coprod : Prop :=
+(inl : Π (A B : C), mono (coprod.inl : A ⟶ A ⨿ B))
+
+variable {C}
+
+@[priority 100]
+instance mono_coprod_of_has_zero_morphisms
+  [has_zero_morphisms C] : mono_coprod C :=
+⟨λ A B, infer_instance⟩
+
+omit hC
+
+lemma mono_sigma_ι_iff_of_is_colimit {J : Type*} (X : J → C) [has_coproduct X]
+  (c : cofan X) (hc : is_colimit c) (j : J) :
+  mono (sigma.ι X j) ↔ mono (c.ι.app (discrete.mk j)) :=
+(morphism_property.respects_iso.monomorphisms C).arrow_iso_iff
+  (arrow.iso_mk' (sigma.ι X j) (c.ι.app (discrete.mk j)) (iso.refl _)
+    (colimit.iso_colimit_cocone ⟨c, hc⟩) (by simp))
+
+include hC
+
+lemma mono_coprod_inl_iff_of_is_colimit {A B : C} (c : binary_cofan A B) (hc : is_colimit c) :
+  mono (coprod.inl : A ⟶ A ⨿ B) ↔ mono c.inl :=
+mono_sigma_ι_iff_of_is_colimit (pair_function A B) c hc walking_pair.left
+
+lemma mono_coprod_inr_iff_of_is_colimit {A B : C} (c : binary_cofan A B) (hc : is_colimit c) :
+  mono (coprod.inr : B ⟶ A ⨿ B) ↔ mono c.inr :=
+mono_sigma_ι_iff_of_is_colimit (pair_function A B) c hc walking_pair.right
+
+omit hC
+
+instance mono_coprod_type : mono_coprod (Type u) :=
+⟨λ A B, begin
+  let c : binary_cofan A B := binary_cofan.mk (sum.inl : A ⟶ A ⊕ B) sum.inr,
```

Comment:
...which would make this proof easier, since you don't need to make the step of transferring from `c` to the chosen colimit